### PR TITLE
RavenDB-19337 Upgrade to Jint 3.0.0-beta-2041

### PR DESCRIPTION
**Pull request by lahma**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-19337

### Additional description

Swithing to official Jint that has API surface for RavenDB customization needs.

### Type of change

- New feature
- Optimization

### How risky is the change?

- Moderate 

### Backward compatibility

- Ensured. Please explain how has it been implemented?

Jint has comprehensive test suite, should not have major regression.

### Is it platform specific issue?

- No

### Documentation update

- No documentation update is needed 

### Testing 

- Should be tested with RavenDB CI test suite and performance test suite

### Is there any existing behavior change of other features due to this change?

- No

### UI work

- No UI work is needed

### CODE REVIEW

File: src/Raven.Server/Documents/Indexes/Static/EsprimaVisitor.cs [hunk 1]
```diff
@@ -562,9 +562,6 @@ public virtual void Visit(Node node)
                 case Nodes.ObjectPattern:
                     VisitObjectPattern(node.As<ObjectPattern>());
                     break;
-                case Nodes.ArrowParameterPlaceHolder:
```
Comment (ayende) [inline src/Raven.Server/Documents/Indexes/Static/EsprimaVisitor.cs:565]: Don't we need that?

File: src/Raven.Server/Documents/Patch/JsBlittableBridge.cs [hunk 2]
```diff
@@ -74,11 +74,11 @@ private void WriteJsonValue(object parent, bool isRoot, string propertyName, obj
...
                     var jsDate = js.AsDate();
-                    if (double.IsNaN(jsDate.PrimitiveValue) ||
-                        jsDate.PrimitiveValue > MaxJsDateMs ||
-                        jsDate.PrimitiveValue < MinJsDateMs)
+                    if (double.IsNaN(jsDate.DateValue) ||
+                        jsDate.DateValue > MaxJsDateMs ||
+                        jsDate.DateValue < MinJsDateMs)
                         // not a valid Date. 'ToDateTime()' will throw
-                        throw new InvalidOperationException($"Invalid '{nameof(DateInstance)}' on property '{propertyName}'. Date value : '{jsDate.PrimitiveValue}'. " +
+                        throw new InvalidOperationException($"Invalid '{nameof(DateInstance)}' on property '{propertyName}'. Date value : '{jsDate.DateValue}'. " +
```
Comment (ayende) [inline src/Raven.Server/Documents/Patch/JsBlittableBridge.cs:81]: What are all the referenced for "use active realms Intrinsics to access well-known object" in the deprecation warnings?

File: src/Raven.Server/Documents/Indexes/Static/EsprimaVisitor.cs (hunk 1 above)
Comment (lahma) [reply src/Raven.Server/Documents/Indexes/Static/EsprimaVisitor.cs:565]: It's internal, during parse-time special placeholder, shouldn't be used by public API.

File: src/Raven.Server/Documents/Patch/JsBlittableBridge.cs (hunk 2 above)
Comment (lahma) [reply src/Raven.Server/Documents/Patch/JsBlittableBridge.cs:81]: Many objects were transferred from `Engine` to separate slot, `Realm`. JS spec has the concept of realms and it also reflects to ability of having `ShadowRealm` separated from the "regual global context". Not sure if `Realm` is the best name or should it just be `Intrinsics` for `Engine` instance. We are trying to follow the spec with naming etc.

**Comment by lahma** [comment]
I'm not quite sure what's the problem with the tests, I cannot catch it with debugger easily.

**Comment by ppekrol** [comment]
Hi @lahma 

We did already a lot of work with new Jint in the https://github.com/ravendb/ravendb/pull/14463
Version there is from Feb 2022.

The v5.4 is an LTS branch, the risk of updating the Jint after the release in LTS build is too high. I think it would be better to push it to 6.0. @lahma Could you do that?

**Comment by lahma** [comment]
@ppekrol I'll try to do that in the near future, if nothing else this diff at least shows where to cheese was moved to...

**Comment by lahma** [comment]
Superseded by https://github.com/ravendb/ravendb/pull/14936

**Pull request closed without merging**